extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -0,0 +1,617 @@
+              partitionOffsets.startingOffset,
+              partition
+          );
+        }
+      } else { // subsequent run, use metadata storage to check consistency
+        partitionOffsets.startingOffset = partitionOffsets.endingOffset;
+        if (storedOffsets.get(partition) == null || storedOffsets.get(partition) != partitionOffsets.endingOffset) {
+          log.error(
+              "endingOffset from metadata storage [%d] does not match expected offset [%d] for partition [%d]",
+              (storedOffsets.get(partition) == null ? -1 : storedOffsets.get(partition)),
+              partitionOffsets.endingOffset,
+              partition
+          );
+        }
+      }
+
+      partitionOffsets.endingOffset = partitionOffsets.startingOffset + messagesPerPartition;
+      partitionOffsets.active = true;
+    }
+  }
+
+  /**
+   * Compares the list of partitions/offsets for taskId from TaskStorage and the data stored in taskGroup.partitions
+   * (which is the current set of partitions/offsets being processed) and returns false if these do not match.
+   */
+  private boolean isTaskCurrent(int taskGroupId, String taskId)
+  {
+    TaskGroup taskGroup = taskGroups.get(taskGroupId);
+    if (!taskStorage.getTask(taskId).isPresent() || taskGroup == null) {
+      return false;
+    }
+
+    KafkaIndexTask task = (KafkaIndexTask) taskStorage.getTask(taskId).get();
+    Map<Integer, Long> taskStartPartitions = task.getIOConfig().getStartPartitions().getPartitionOffsetMap();
+    Map<Integer, Long> taskEndPartitions = task.getIOConfig().getEndPartitions().getPartitionOffsetMap();
+
+    for (Integer partition : taskStartPartitions.keySet()) {
+      if (taskGroup.partitions.get(partition) == null
+          || taskGroup.partitions.get(partition).startingOffset != taskStartPartitions.get(partition)
+          || taskGroup.partitions.get(partition).endingOffset != taskEndPartitions.get(partition)) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  private void killRunningTasksInGroup(int taskGroupId)
+  {
+    if (taskMaster.getTaskQueue().isPresent()) {
+      Map<String, Optional<TaskStatus>> taskStatuses = taskGroups.get(taskGroupId).taskStatuses;
+      for (String taskId : taskStatuses.keySet()) {
+        if (taskStatuses.get(taskId).isPresent() && !taskStatuses.get(taskId).get().isComplete()) {
+          taskMaster.getTaskQueue().get().shutdown(taskId);
+        }
+      }
+    } else {
+      log.error("Failed to get task queue because I'm not the leader!");
+    }
+  }
+
+  private void killUnownedKafkaTasks()
+  {
+    if (taskMaster.getTaskQueue().isPresent()) {
+      List<Task> tasks = taskStorage.getActiveTasks();
+      for (Task task : tasks) {
+        if (task instanceof KafkaIndexTask
+            && dataSource.equals(task.getDataSource())
+            && !supervisorRunId.equals(task.getContextValue(SUPERVISOR_CONTEXT_KEY))) {
```

Comment:
Hmm, this looks like it's going to kill any task that wasn't started by us. That's more aggressive than we really want, since it means tasks will be dying during a normal restart of overlords. Ideally the tasks shouldn't die.

So it'd be better to let the tasks live if they're doing something we approve of. I guess we can talk about what exactly that means but I think a place to start thinking about that would be:

- The replication stuff works properly if the tasks create the same segments. So that at least means we want the same dataSchema, startPartitions, endPartitions, and maxRowsPerSegment. Possibly other stuff too.
- If our instances of those things are the same as what's already running, we shouldn't need to kill what's running.
- If our instances of those things are *not* the same as what's already running, we could kill what's running, or we could let it stay and just adjust how we replace things (i.e. replace dead tasks in existing task groups with tasks that are "compatible" with existing ones, rather than using our templates for new tasks)

This may be something we want to address now or may be something we want to file a follow-on issue for…